### PR TITLE
profiler: collect profiles concurrently

### PR DESCRIPTION
**Pull request by nsrip-dd**
The upcoming C allocation profile will work by starting a profile,
waiting for the profile period to elapse, then stopping and collecting
the output. This won't work if the profiles are collected synchronously
since the CPU profile is also collected this way, meaning there won't be
enough time in a single profile period to collect both. So the profiles
need to be collected concurrently.

### CODE REVIEW

File: profiler/profiler.go [hunk 1]
```diff
@@ -211,6 +212,12 @@ func (p *profiler) run() {
 // an item.
 func (p *profiler) collect(ticker <-chan time.Time) {
 	defer close(p.out)
+	var (
+		// mu guards completed
+		mu        sync.Mutex
+		completed []*profile
```
Comment (felixge) [inline profiler/profiler.go:218]: NIT: mu might not be needed if pre-allocating `completed` to the right size and putting each profile in its own slot when its done.

File: profiler/profiler.go [hunk 2]
```diff
@@ -226,15 +233,22 @@ func (p *profiler) collect(ticker <-chan time.Time) {
...
+				go func(t ProfileType) {
+					defer wg.Done()
+					profs, err := p.runProfile(t)
+					if err != nil {
+						log.Error("Error getting %s profile: %v; skipping.", t, err)
+						p.cfg.statsd.Count("datadog.profiler.go.collect_error", 1, append(p.cfg.tags, t.Tag()), 1)
+					}
+					mu.Lock()
+					defer mu.Unlock()
+					completed = append(completed, profs...)
```
Comment (felixge) [inline profiler/profiler.go:246]: 🐞 `completed` growing unbound, not getting reset (found by nick)

testing nit: would be nice if we had a test that does multiple rounds of collect/upload to catch these kind of bugs in the future.

**Review by felixge — CHANGES_REQUESTED** [review]
LGTM in general, but see comments.

File: profiler/profiler.go (hunk 2 above)
Comment (nsrip-dd) [reply profiler/profiler.go:247]: Done, updated `TestAllUploaded` to check the second batch of profiles, which failed before fixing this bug.

File: profiler/profiler.go (hunk 1 above)
Comment (nsrip-dd) [reply profiler/profiler.go:218]: I think I'll keep it this way for now for this reason: What if one of the profiles fails? Do I put `nil` at the index where the profile would go and check for nil before adding it to the batch at the end? Also if we go back to a state where some profile types produce multiple profiles (like delta + non-delta) then we'd need to do more bookkeeping to figure out the offset in the array where the profiles should go for each type.

**Review by felixge — APPROVED** [review]
LGTM 🙏

File: profiler/profiler_test.go [hunk 3]
```diff
@@ -344,17 +348,31 @@ func TestAllUploaded(t *testing.T) {
...
+			default:
+			}
+		}()
 		_, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
 		if err != nil {
 			t.Fatalf("bad media type: %s", err)
 			return
 		}
 		mr := multipart.NewReader(r.Body, params["boundary"])
+		profiles = profiles[:0]
```
[the remark this answers is not in the document]
Comment (nsrip-dd) [reply profiler/profiler_test.go:375]: 👍 I was thinking we could verify the profilers in the handler (see this cleanup PR: https://github.com/DataDog/dd-trace-go/pull/1267) which would mean all uploads get checked.

[the remark this answers is not in the document]
Comment (felixge) [reply profiler/profiler_test.go:375]: Should I review https://github.com/DataDog/dd-trace-go/pull/1267 now?

[the remark this answers is not in the document]
Comment (nsrip-dd) [reply profiler/profiler_test.go:375]: I'll ping you when it's ready for review since I have to fix some merge conflicts for that PR (caused by this one)

**Pull request merged**